File: sourceCode/algorithm_python/1967_트리의지름.py
import sys
from collections import deque
input = sys.stdin.readline


def bfs(start):
	visited = [0] * (N + 1)
	visited[start] = 1
	queue = deque()
	queue.append(start)
	distance = [0] * (N + 1)
	max_distance = 0
	max_node = start
	while queue:
		cur = queue.popleft()

		for nxt, n_d in graph[cur]:
			if visited[nxt] == 0:
				visited[nxt] = 1
				distance[nxt] = distance[cur] + n_d
				if max_distance < distance[nxt]:
					max_distance = distance[nxt]
					max_node = nxt
				queue.append(nxt)
	return (max_distance, max_node)

if __name__=='__main__':
	N = int(input())
	graph = [[] for _ in range(N + 1)]
	for _ in range(N-1):
		s, e, d = map(int, input().split())
		graph[s].append((e, d))
		graph[e].append((s, d))
	distance, node = bfs(1)
	distance, node = bfs(node)
	print(distance)

File: sourceCode/algorithm_Cplusplus/1967_트리의지름.cpp
#include <iostream>
#include <vector>
using namespace std;
int N;
vector<vector<pair<int, int>>> graph = vector<vector<pair<int, int>>>();

pair<int, int> bfs(int start)
{
	int visited[N + 1] = {};
	int distance[N + 1] = {};
	int max_distance = 0;
	int max_node = start;
	visited[start] = 1;
	vector<int> queue = vector<int>();
	queue.push_back(start);
	while (!queue.empty())
	{
		int cur = queue.front();
		queue.erase(queue.begin());
		for (pair<int, int> nxt : graph[cur])
		{
			if (visited[nxt.first] == 0)
			{
				visited[nxt.first] = 1;
				distance[nxt.first] = distance[cur] + nxt.second;
				if (max_distance < distance[nxt.first])
				{
					max_distance = distance[nxt.first];
					max_node = nxt.first;
				}
				queue.push_back(nxt.first);
			}
		}
	}
	return pair(max_node, max_distance);
}

int main(void)
{
	cin >> N;
	for (int i = 0; i < N + 1; i++)
		graph.push_back(vector<pair<int, int>>());
	for (int i = 0; i < N - 1; i++)
	{
		int s, e, d;
		cin >> s >> e >> d;
		graph[s].push_back(pair(e, d));
		graph[e].push_back(pair(s, d));
	}
	pair<int, int> result;
	result = bfs(1);
	result = bfs(result.first);
	cout << result.second;
	return 0;
}